src/lib/core/data-table/data-table.ts

```diff
@@ -142,40 +175,54 @@ export class CdkTable<T> implements CollectionViewer {
       this._columnDefinitionsByName.set(columnDef.name, columnDef);
     });
 
-    // Get and merge the streams for column changes made to the row defs
-    const rowDefs = [...this._rowDefinitions.toArray(), this._headerDefinition];
-    const columnChangeStreams =
-        rowDefs.map((rowDef: BaseRowDef) => rowDef.columnsChange);
-    this._columnsChange = Observable.merge(...columnChangeStreams);
+    // Re-render the rows if any of their columns change.
+    // TODO(andrewseguin): Determine how to only re-render the rows that have their columns changed.
+    const rowColumnsChange =
+        this._rowDefinitions.toArray().map((rowDef: BaseRowDef) => rowDef.columnsChange);
+    Observable.merge(...rowColumnsChange)
+        .takeUntil(this._onDestroy)
+        .subscribe(() => {
+          // Reset the data to an empty array so that renderRowChanges will re-render all new rows.
+          this._rowPlaceholder.viewContainer.clear();
+          this._dataDiffer.diff([]);
+          this._renderRowChanges();
+        });
+
+    // Re-render the header row if the columns change
+    this._headerDefinition.columnsChange
+        .takeUntil(this._onDestroy)
+        .subscribe(() => {
+          this._headerRowPlaceholder.viewContainer.clear();
+          this._renderHeaderRow();
+        });
   }
 
   ngAfterViewInit() {
-    this.renderHeaderRow();
+    this._renderHeaderRow();
 
-    // Re-render the header row if the columns changed.
-    this._columnsChange.subscribe(() => {
-      this._headerRowPlaceholder.viewContainer.clear();
-      this.renderHeaderRow();
+    if (this.dataSource) {
+      this._observeRenderChanges();
+    }
 
-      // Reset the data to an empty array so that renderRowChanges will re-render all new rows.
-      this._rowPlaceholder.viewContainer.clear();
-      this._dataDiffer.diff([]);
-    });
+    this._isViewInitialized = true;
+  }
 
-    // TODO(andrewseguin): If the data source is not
-    //   present after view init, connect it when it is defined.
-    // TODO(andrewseguin): Unsubscribe from this on destroy.
-    const streams = [this.dataSource.connect(this), this._columnsChange];
-    Observable.combineLatest(streams).subscribe(([rowsData]) => {
-      this.renderRowChanges(rowsData);
-    });
+  /** Set up a subscription for the data provided by the data source. */
+  _observeRenderChanges() {
+    this._renderChangeSubscription = this.dataSource.connect(this)
+        .takeUntil(this._onDestroy)
+        .subscribe(data => {
+          this._data = data;
+          this._renderRowChanges();
+        });
   }
 
   /**
    * Create the embedded view for the header template and place it in the header row view container.
    */
-  renderHeaderRow() {
-    const cells = this.getHeaderCellTemplatesForRow(this._headerDefinition);
+  _renderHeaderRow() {
```

Comment:
Can this be marked private? Or is it used somewhere else? Ditto to some of the other methods in this class.